daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala

```diff
@@ -167,79 +167,83 @@ class ContractStateMachine[Nid](mode: ContractKeyUniquenessMode) {
-                // An archive can only mark a key as inactive
-                // if it was brought into scope before.
-                case Some(KeyActive(cid)) if cid == targetId =>
-                  consumedState.copy(keys = updatedKeys)
-                // If the key was not in scope or mapped to a different cid, we don’t change keys. Instead we will do
-                // an activeness check when looking it up later.
-                case _ => consumedState
+    ): Either[InconsistentContractKey, State] = {
+      for {
+        state <-
+          if (byKey || !mode.byKeyOnly)
+            assertKeyMapping(templateId, targetId, mbKey)
+          else Right(this)
+      } yield {
+        if (consuming) {
+          val consumedState = state.activeState.consume(targetId, nodeId)
+          val newActiveState = mbKey match {
+            case Some(kWithM) =>
+              val gkey = GlobalKey(templateId, kWithM.key)
+              val keys = consumedState.keys
+              val updatedKeys = keys.updated(gkey, KeyInactive)
+
+              // If the key was brought in scope before, we must update `keys`
+              // independently of whether this exercise is by-key because it affects later key lookups.
+              if (mode.byKeyOnly) {
+                keys.get(gkey).orElse(state.lookupActiveGlobalKeyInput(gkey)) match {
+                  // An archive can only mark a key as inactive
+                  // if it was brought into scope before.
+                  case Some(KeyActive(cid)) if cid == targetId =>
+                    consumedState.copy(keys = updatedKeys)
+                  // If the key was not in scope or mapped to a different cid, we don’t change keys. Instead we will do
+                  // an activeness check when looking it up later.
+                  case _ => consumedState
+                }
+              } else {
+                consumedState.copy(keys = updatedKeys)
               }
-            } else {
-              consumedState.copy(keys = updatedKeys)
-            }
-          case None => consumedState
-        }
-        this.copy(activeState = newActiveState)
-      } else this
+            case None => consumedState
+          }
+          state.copy(activeState = newActiveState)
+        } else state
+      }
     }
 
     def handleLookup(lookup: Node.LookupByKey): Either[KeyInputError, State] = {
       // If the key has not yet been resolved, we use the resolution from the lookup node,
       // but this only makes sense if `activeState.keys` is updated by every node and not only by by-key nodes.
       require(!mode.byKeyOnly, "This method can only be used if all key nodes are considered")
-      handleLookupWith(lookup, lookup.result)
+      visitLookup(lookup.templateId, lookup.key.key, lookup.result).left.map(Left(_))
     }
 
-    def handleLookupWith(
-        lookup: Node.LookupByKey,
+    private[lf] def visitLookup(
```

Comment:
Don't make this private. Canton needs to call this method with an input that is not `keyInput` for now (until we define a sane semantics for non-UCK mode).
```suggestion
    def visitLookup(
```

Reply:
oh sorry I missed the renaming here. I would suggest to keep calling it `handleLookupWith` with the same type signature. Then we have the invariant that `handle*` methods are public, `visit` methods are Daml Engine only.